src/components/LinkModal/LinkModal.js

```diff
@@ -0,0 +1,253 @@
+    selectors.getCurrentPage(state),
+    selectors.getSelectedTab(state, 'linkModal'),
+  ]);
+  const [t] = useTranslation();
+  const dispatch = useDispatch();
+
+  const urlInput = React.createRef();
+  const pageNumberInput = React.createRef();
+
+  const [dropdownNumbers, setDropdownNumbers] = useState([]);
+  const [url, setURL] = useState('');
+  const [pageNumber, setPageNumber] = useState(1);
+
+  const closeModal = () => {
+    dispatch(actions.closeElement('linkModal'));
+    setURL('');
+    setPageNumber(1);
+    core.setToolMode(defaultTool);
+  };
+
+  const createLink = () => {
+    const linksResults = [];
+
+    const quads = core.getSelectedTextQuads();
+    const selectedAnnotations = core.getSelectedAnnotations();
+
+    const currPageQuads = quads[currentPage - 1];
+
+    if (currPageQuads) {
+      const selectedText = core.getSelectedText();
+      currPageQuads.forEach(quad => {
+        const link = new Annotations.Link();
+        link.PageNumber = currentPage;
+        link.StrokeColor = new Annotations.Color(0, 165, 228);
+        link.StrokeStyle = 'underline';
+        link.StrokeThickness = 2;
+        link.Author = core.getCurrentUser();
+        link.Subject = 'Link';
+        link.X = Math.min(quad.x1, quad.x3);
+        link.Y = Math.min(quad.y1, quad.y3);
+        link.Width = Math.abs(quad.x1 - quad.x3);
+        link.Height = Math.abs(quad.y1 - quad.y3);
+        linksResults.push(link);
+      });
+      createHighlightAnnot(linksResults, currPageQuads, selectedText);
+    }
+
+    if (selectedAnnotations && selectedAnnotations.length === 1) {
+      const link = new Annotations.Link();
+      link.PageNumber = currentPage;
+      link.StrokeColor = new Annotations.Color(0, 165, 228);
+      link.StrokeStyle = 'underline';
+      link.StrokeThickness = 2;
+      link.Author = core.getCurrentUser();
+      link.Subject = 'Link';
+      link.X = selectedAnnotations[0].X;
+      link.Y = selectedAnnotations[0].Y;
+      link.Width = selectedAnnotations[0].Width;
+      link.Height = selectedAnnotations[0].Height;
+      linksResults.push(link);
+      selectedAnnotations[0].associateLink([link.Id]);
+    }
+
+    return linksResults;
+  };
+
+  const createHighlightAnnot = async(linkAnnotArray, quads, text) => {
+    const linkAnnot = linkAnnotArray[0];
+    const highlight = new Annotations.TextHighlightAnnotation();
+    highlight.PageNumber = linkAnnot.PageNumber;
+    highlight.X = linkAnnot.X;
+    highlight.Y = linkAnnot.Y;
+    highlight.Width = linkAnnot.Width;
+    highlight.Height = linkAnnot.Height;
+    highlight.StrokeColor = new Annotations.Color(0, 0, 0, 0);
+    highlight.Quads = quads;
+    highlight.Author = core.getCurrentUser();
+    highlight.setContents(text);
+
+    let linkAnnotIdArray = [];
+    linkAnnotArray.forEach(link => {
+      linkAnnotIdArray.push(link.Id);
+    });
+    highlight.associateLink(linkAnnotIdArray);
+
+    await core.addAnnotations([highlight]);
```

Comment:
Does `addAnnotations` return a promise? I'm not sure if we need to await this function and `drawAnnotationsFromList`

Reply:
await was necessary on that one, API doc didn't say that but after I removed it, my link annots stopped rendering.

Reply:
I can reproduce the issue you were having. I tried to remove the `await`s and can only see the link annotations appear after I zoom in. I still didn't figure out why but I'm a bit concerned about using `await` to make the code work since apparently  `annotManager.addAnnotations` is synchronous and doesn't return a promise.

After checking the code it seems that link annotations is a HTML annotation, and to draw it we need to call `annotManager.drawAnnotations` with the third argument being true. The third argument is `majorRedraw` and when it's true it will redraw HTML annotations such as link annotations. 

However even if I ask annotManager to major redraw all the annotations in the page I still didn't see the link annotations appear... I found that If I don't draw the highlight annotations before drawing the link annotations everything will be good. This is the code I have

```javascript
const createHighlightAnnot = (linkAnnotArray, quads, text) => {
  const linkAnnot = linkAnnotArray[0];
  const highlight = new Annotations.TextHighlightAnnotation();
  highlight.PageNumber = linkAnnot.PageNumber;
  highlight.X = linkAnnot.X;
  highlight.Y = linkAnnot.Y;
  highlight.Width = linkAnnot.Width;
  highlight.Height = linkAnnot.Height;
  highlight.StrokeColor = new Annotations.Color(0, 0, 0, 0);
  highlight.Quads = quads;
  highlight.Author = core.getCurrentUser();
  highlight.setContents(text);

  const linkAnnotIdArray = linkAnnotArray.map(link => link.Id);
  highlight.associateLink(linkAnnotIdArray);

  core.addAnnotations([highlight]);
  // core.drawAnnotationsFromList([highlight]);
};

const addURLLink = () => {
  const links = createLink();

  const action = new window.Actions.URI({ uri: url });
  links.forEach(link => {
    link.addAction('U', action);
    core.addAnnotations([link]);
    core.drawAnnotations(link.PageNumber, null, true);
  });

  closeModal();
};
```

Not sure if this is better though...

Reply:
Resolved, thx Z.